### PR TITLE
Add "flutter downgrade" command

### LINKED ISSUE #40691: Write an integration test that exercises `flutter upgrade`
We should write an integration test that exercises the `flutter upgrade` flow, so that we know when we break the upgrade flow.

This should be possible via fudging the git history. NOTE: that in our code, we hardcode the upstream repo.

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

In cases where a flutter upgrade forces a user to contend with breaking changes they may not wish to deal with, the flutter downgrade command takes them to the last version they were at for a particular channel. It does so by storing a version per official channel (master, dev, beta, stable) when an upgrade is performed. The mechanics of downgrade are intentionally simpler - it does a hard reset to the particular revision and then checks out the channel. This should be safe since we'll maintain the versions per channel

@zanderso for initial thoughts on the mechanics here

EDIT: more context, we've got a lot of crashes from the flutter version command the last time we did a release, current working theory is that some people try to go back to the last version. This should make the easier, though we might also need to think about how we advertise it

Fixes https://github.com/flutter/flutter/issues/40691

### CODE REVIEW

**Comment by jonahwilliams** [comment]
Also @tvolkert and @christopherfujino

File: packages/flutter_tools/test/commands.shard/hermetic/downgrade_test.dart [hunk 1]
```diff
@@ -0,0 +1,169 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
```
Comment (tvolkert) [inline packages/flutter_tools/test/commands.shard/hermetic/downgrade_test.dart]: Can you add a test of downgrading to a hotfix (branched) version, then upgrading again?

File: packages/flutter_tools/lib/src/commands/downgrade.dart [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+/// The flutter downgrade command returns the SDK to the last recorded version
+/// for a particular branch.
+///
+/// For example, suppose a user on the beta channel upgrades from 1.2.3 to 1.4.6.
+/// The tool will record that 1.2.3 was the last active beta channel in the
+/// persistent tool state. If the user is still on the beta channel and runs
+/// flutter downgrade, this will take the user back to 1.2.3. They will not be
+/// able to downgrade again, since the tool only records one prior version.
+/// Additionally, if they had switch channels to stable before trying to downgrade,
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/commands/downgrade.dart]: sp: `"switch"` -> `"switched"`

File: packages/flutter_tools/lib/src/persistent_tool_state.dart [hunk 3]
```diff
@@ -63,6 +72,12 @@ class _DefaultPersistentToolState implements PersistentToolState {
 
   static const String _kFileName = '.flutter_tool_state';
   static const String _kRedisplayWelcomeMessage = 'redisplay-welcome-message';
+  static const List<String> _lastActiveVersion = <String>[
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/persistent_tool_state.dart]: What if this was a `Map<Channel, String>`, and then:

```
  String _versionKeyFor(Channel channel) => _lastActiveVersion[channel];
```

File: packages/flutter_tools/lib/src/version.dart [hunk 4]
```diff
@@ -15,10 +15,36 @@ import 'cache.dart';
 import 'convert.dart';
 import 'globals.dart' as globals;
 
+/// The names of each channel/branch ordered by stability.
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/version.dart]: -> in order of increasing stability.

**Review by zanderso — COMMENTED** [review]
Three things:
- integration tests
- full git hash in .flutter_tool_state
- add a message to `flutter version` pointing to `flutter downgrade` and warning about detached HEAD state, etc.

**Comment by jonahwilliams** [comment]
RE: the downgrade warning message. Do you think we should include full instructions on getting back to a known state (what are they...?) or is the message itself enough

**Comment by jonahwilliams** [comment]
I've added an integration test that exercises both upgrade and downgrade (from/to a hotfix). This requires a secondary checkout and the ability to override the working directory

File: packages/flutter_tools/test/commands.shard/hermetic/downgrade_test.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/test/commands.shard/hermetic/downgrade_test.dart]: How about:

1. Start at specified hotfix version on stable.
2. Upgrade to newest stable.
3. Downgrade back to original hotfix?

File: packages/flutter_tools/lib/src/commands/downgrade.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/commands/downgrade.dart]: Done

File: packages/flutter_tools/lib/src/persistent_tool_state.dart (hunk 3 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/persistent_tool_state.dart]: Done

File: packages/flutter_tools/lib/src/commands/downgrade.dart [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+  AnsiTerminal _terminal;
+  FlutterVersion _flutterVersion;
+  PersistentToolState _persistentToolState;
+  ProcessUtils _processUtils;
+  ProcessManager _processManager;
+  Logger _logger;
+  Stdio _stdio;
+
+  @override
+  String get description => 'downgrade Flutter to the last active version for the current channel.';
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/commands/downgrade.dart]: Done

File: packages/flutter_tools/lib/src/commands/downgrade.dart [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+    _processManager ??= globals.processManager;
+    _processUtils ??= ProcessUtils(processManager: _processManager, logger: _logger);
+    _stdio ??= globals.stdio;
+
+    final String currentChannel = _flutterVersion.channel;
+    final Channel channel = getChannelForName(currentChannel);
+    if (channel == null) {
+      throwToolExit(
+        'Flutter is not currently on a known channel. Use "flutter channel <name>" '
+        'to switch to an official channel.');
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/commands/downgrade.dart]: Done

File: packages/flutter_tools/lib/src/commands/downgrade.dart [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+    }
+    final GitTagVersion lastFlutterVesion = _persistentToolState.lastActiveVersion(channel);
+    final GitTagVersion currentFlutterVersion = GitTagVersion.determine(_processUtils);
+    if (lastFlutterVesion == null ||
+        currentFlutterVersion.isEquivalent(lastFlutterVesion) ||
+        lastFlutterVesion.isEquivalent(const GitTagVersion.unknown())) { // condition detections unknown version.
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/commands/downgrade.dart]: Done

File: packages/flutter_tools/lib/src/version.dart (hunk 4 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/version.dart]: Done

File: packages/flutter_tools/lib/src/commands/downgrade.dart [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+      throwToolExit(
+        'There is no previously recorded version for channel "$currentChannel."'
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/commands/downgrade.dart]: Done

**Comment by jonahwilliams** [comment]
This is currently hitting filename too long errors on windows

File: packages/flutter_tools/lib/src/commands/version.dart [hunk 9]
```diff
@@ -57,8 +57,26 @@ class VersionCommand extends FlutterCommand {
     final List<String> tags = await getTags();
     if (argResults.rest.isEmpty) {
       tags.forEach(globals.printStatus);
-      return const FlutterCommandResult(ExitStatus.success);
+      return FlutterCommandResult.success();
     }
+
+    globals.printStatus(
+      'Warning: "flutter version" will leave the SDK in a detached HEAD state. '
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/commands/version.dart]: Consider adding an ascii border as in the welcome message in usage.dart.

File: packages/flutter_tools/lib/src/version.dart [hunk 10]
```diff
@@ -15,14 +15,41 @@ import 'cache.dart';
...
+  }
+  return null;
+}
+
 class FlutterVersion {
-  FlutterVersion([this._clock = const SystemClock()]) {
-    _frameworkRevision = _runGit(gitLog(<String>['-n', '1', '--pretty=format:%H']).join(' '));
-    _gitTagVersion = GitTagVersion.determine();
+  FlutterVersion([this._clock = const SystemClock(), this._workingDirectory]) {
+    _frameworkRevision = _runGit(gitLog(<String>['-n', '1', '--pretty=format:%H']).join(' '), processUtils, _workingDirectory);
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/version.dart]: Split some of these now long parameter lists:
```dart
_frameworkRevision = _runGit(
  gitLog(<String>['-n', '1', '--pretty=format:%H']).join(' '),
  processUtils, _workingDirectory,
);
```

File: packages/flutter_tools/test/commands.shard/permeable/upgrade_test.dart [hunk 11]
```diff
@@ -57,6 +57,7 @@ void main() {
 
     testUsingContext('throws on unknown tag, official branch,  noforce', () async {
       final Future<FlutterCommandResult> result = fakeCommandRunner.runCommand(
+        false,
```
Comment (zanderso) [inline packages/flutter_tools/test/commands.shard/permeable/upgrade_test.dart]: Named parameters could help with readability here.

**Review by zanderso — APPROVED** [review]
lgtm w/ nits

File: packages/flutter_tools/test/commands.shard/permeable/upgrade_test.dart (hunk 11 above)
Comment (jonahwilliams) [reply packages/flutter_tools/test/commands.shard/permeable/upgrade_test.dart]: Done

File: packages/flutter_tools/lib/src/version.dart (hunk 10 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/version.dart]: Done

File: packages/flutter_tools/lib/src/commands/version.dart (hunk 9 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/commands/version.dart]: Done

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/commands/version.dart]: I verified the borders lined up in my own terminal, I think that should be sufficient unless someone isn't using a monospace font, right?

Comment (zanderso) [reply packages/flutter_tools/lib/src/commands/version.dart]: If someone isn't using a monospace font in a terminal, then there's nothing we can do for them.

File: packages/flutter_tools/lib/src/commands/downgrade.dart [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+       _flutterVersion = flutterVersion,
+       _persistentToolState = persistentToolState,
+       _processManager = processManager,
+       _stdio = stdio,
+       _logger = logger,
+       _fileSystem = fileSystem {
+    argParser.addOption(
+      'working-directory',
+      hide: true,
+      help: 'Override the downgrade working directoy for integration testing.'
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/commands/downgrade.dart]: sp: `directoy` -> `directory

**Review by christopherfujino — APPROVED** [review]
LGTM

**Comment by Ahmadre** [comment]
Thanks @jonahwilliams :)

**Pull request merged**